Core/StitchCoreSDK/Sources/StitchCoreSDK/Auth/Internal/CoreStitchAuth+OtherAuthActions.swift

```diff
@@ -0,0 +1,177 @@
+import Foundation
+import MongoSwift
+
+extension CoreStitchAuth {
+    /**
+     * Logs out the current user, and clears authentication state from this `CoreStitchAuth` as well as underlying
+     * storage. Blocks the current thread until the request is completed. If the logout request fails, this method will
+     * still clear local authentication state.
+     */
+    public func logoutInternal(withId userId: String?) throws {
+        objc_sync_enter(authOperationLock)
+        defer { objc_sync_exit(authOperationLock) }
+
+        // Unwrap the userId or use the active user
+        guard let userId = userId != nil ? userId : activeUserAuthInfo?.userID else {
+            return
+        }
+
+        // Get the user objects if it exists --> otherwise throw
+        guard let index = loggedInUsersAuthInfo.firstIndex(where: {$0.userID == userId}) else {
+            throw StitchError.serviceError(
+                withMessage: "User with id: \(userId) not found",
+                withServiceErrorCode: .userNotFound)
+        }
+        let authInfo = loggedInUsersAuthInfo[index]
+
+        // If the user is not logged in --> return
+        guard authInfo.isLoggedIn else { return }
+
+        // If it is an AnonymousUser remove the user altogether from the list
+        // Otherwise set the authInfo to have no access token or refresh token
+        var newAuthInfo = StoreAuthInfo.init(withAuthInfo: authInfo, withLogout: true).toAuthInfo
+        newAuthInfo = StoreAuthInfo.init(withAuthInfo: newAuthInfo, withNewTime: true).toAuthInfo
+
+        if authInfo.loggedInProviderType == .anonymous {
+            loggedInUsersAuthInfo.remove(at: index)
```

Comment:
[code organization] IMO it would be better to call `removeUserInternal` here. This way once we add in the hooks for the new auth listener, you don't have to look for all the places where a user can be removed